### ISSUE TITLE
software issue

### ISSUE THREAD

**Issue by aldendirks** [SysBioChalmers/HGTphyloDetect#11]
This pipeline was not ready for deployment to the general public and saying it is easy to use is false advertising. I encountered several issues, many of which I fixed on my own, but ultimately it is taking too much time to get this pipeline to work so I’m moving on. Below is a list of issues that I encountered. If they were fixed in the next version, I would try this pipeline again and maybe recommend it to others. But until then, to those thinking about trying it out, do not attempt to use this software unless you want to take the general ideas and basically write your own pipeline.  

EDIT: [AvP](https://github.com/GDKO/AvP) worked beautifully first try. 

### Issue 1: Input protein accessions need to match a GenBank accession number

This issue has been brought up by others. If you are using unpublished protein sequences you will need to modify the “HGT_workflow” Python scripts to manually specify your taxonomic group for alien index calculations.

### Issue 2:  BLASTP set to “remote”

Not a problem in itself, but after a few BLASTs NCBI reports a message that the IP address will be penalized for overuse. You get this same message if you BLAST too many sequences on their web server GUI. An option to specify remote or local database search would be nice. You can go into the BLASTP Python script and remove the remote flag. 

### Issue 3: NCBI taxonomy parsing errors spoil the run 

Not every taxon has an NCBI tax ID at every taxonomic level. In the HGT output Python code, if one of the requested tax IDs is missing the run fails. I added some “try, except” blocks here so that the run wouldn’t crash if “taxid2name” failed. 

### Issue 4: After that, still not every gene is reported in the output

I don’t know why, but the log files to which I directed the code show that multiple HGT events were found and alien indexes calculated but they are ultimately not being written to the output TSV. This is a problem for non-HGT genes too. Basically a large number of the genes are not being reported to the output TSV and I don’t know why. Perhaps because of “taxid2name” and “ranks2lineage” function errors that the script can’t handle. 

### Issue 5: No connection between two modules of pipeline

This issue is another that has already been brought up. The BLASTP output format required for the alien index calculator is different than the format required for the phylogenetic steps, and the pipeline does not have a way to link these parts. One could match the BLASTP outputs for the two steps but would need change one of the Python scripts so that the ncbi_parse function indexing (accession number and percent similarity or evalue) is correct. In general, the output of the alien index calculator part and the phylogenetic part do not link up such that you need to write your own code anyway to go through the TSV output and select the genes that have been identified as HGT. As a result, it was easier for me to run BLASTP twice: first time for all the genes of interest with the output format compatible for the alien index calculator; second time just for the HGT genes with the output format compatible for the phylogenetic steps. 

### Issue 6: Failure to generate homologs FASTA file for phylogenetics

But even then the homologs FASTA file fails to generate for some genes even though the log file looks OK (the accessions are being reported with the correct taxonomy designation). I don't know why some genes work and some don't. The one that did work was clearly not HGT in the end. 

### Issue 7: Script requires very particular directory structure

The pipeline is not written such that the path to your input and path to your output can be specified. Instead, you have to recreate a directory tree that matches the particular way the pipeline was run for the paper in which it was published. 

### Issue 8: It's slow

The pipeline is slow to run compared to AvP, which completed the calculations very quickly. 

### In summary

All in all, this pipeline was generated for a particular use case and then published as a general tool but it has not been coded or tested to work on datasets that are not the original one. Even then there are steps in the pipeline that are missing (connecting the alien index calculation and phylogenetic portion) that you need to figure out on your own.

**Comment by le-yuan** [comment]
Hi,

I am sorry that you have bad feelings about our computational toolbox HGTphyloDetect. But some of your comments are too harsh for us.

Firstly, let's check the title of your comments: "**Don’t waste your time, software not usable in current form**". If you check the recent published paper [Long-read genome sequencing provides novel insights into the harmful algal bloom species _Prymnesium parvum_](https://www.sciencedirect.com/science/article/pii/S004896972306669X) in the journal [Science of The Total Environment](https://www.sciencedirect.com/journal/science-of-the-total-environment) (Impact factor is about 10), you could know that there are already some scientists using our toolbox HGTphyloDetect in their own research.

Then, I also want to explain something about your listed issues (not one by one). For example:

Currently, the input protein accessions should come from the widely used GenBank database. If you check our pipeline, the protein id is important for us because we need to retrieve taxonomy information from there. And the GenBank database has contained **247777761 specific sequences (version 258, Oct 2023)** maintained by NIH, which is a super cool resource updated since Dec 1982. I believe the current version can satisfy many many users in the community.

Right now, Blastp was run by using NCBI remote blastp function. As you may know, there are mainly two approaches that can be used to run Blastp. The first one is NCBI Blastp, users don't need to download the large NCBI database if they choose the remote Blastp, it is convenient. The second one is Diamond. While it is very fast, users need to download very large database before using the software. Actually, I also tried this option before, I need to spend **more than 24 hours** to download database if I remember correctly.

For the taxonomy information parsing, I used the **toolkit ETE** (http://etetoolkit.org/docs/latest/tutorial/tutorial_ncbitaxonomy.html), which is very fast and very convenient! Besides, I also added the "try, except" block in the script to make sure the not found entries would not throw an error.

Although the software in running HGT prediction is not very quick, the **accuracy and sensitivity are quite good by using some specific species as case studies**. And we have made some descriptions in the discussion part in our published paper.

As you listed, particular directory structure is needed in HGTphyloDetect. However, we also provided **detailed documentation** about it. Especially, we provided detailed examples about how to run those steps one by one and what is the output.

Let's look forward to checking the future version of our software, hopefully you and many other people would like it! Also **thanks for your feedback**!
